Gi 251 mass tooltips for buy menu

traunts opened the pull request:
Closes #251 . Adds property info to the description tooltip for items in the shop list and cart of the BuyMenu GUI. 
Craft now display max carriable mass, and maximum passenger slots.
Actors now display their mass, and passenger slots required if greater than 1.
All other MoveableObjects now display their mass.

traunts commented:
Concerned about the correctness of the float->int conversion for masses. That method rounded values correctly, in that they match the max mass of craft (drop pod not included). Originally I was using `floor/ceil`, had troubles making it readable and avoid rounding errors. Thoughts appreciated.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1208,20 +1208,45 @@ void BuyMenuGUI::Update()
-        {
-            const DataModule *pModule = g_PresetMan.GetDataModule(pItem->m_ExtraIndex);
-            if (pModule && !pModule->GetDescription().empty())
+            const Entity* currentItem = pItem->m_pEntity;
```
Formatting - for pointer and reference variables we always go with `Type *variable` or `Type &variable`. For pointer and reference methods, it looks like `Type * method()` or `Type & method()`.

Applies throughout your changes.

traunts replied:
What's the correct case for situations like casting? I'm currently using `dymanic_cast<const Type*>`, for example.

garethyr replied:
You'll wanna do `dynamic_cast<Type *>(thing)` or `dynamic_cast<const Type *>(thing)`

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1208,20 +1208,45 @@ void BuyMenuGUI::Update()
-            const DataModule *pModule = g_PresetMan.GetDataModule(pItem->m_ExtraIndex);
-            if (pModule && !pModule->GetDescription().empty())
+            const Entity* currentItem = pItem->m_pEntity;
+            const ACraft* pItemAsCraft = dynamic_cast<const ACraft*>(currentItem);
```
Naming, we only include prefixes for member variables, and never the `p` pointer prefix. Intellisense and context can tell you if a variable is a pointer just fine.

Applies throughout your changes.

traunts replied:
Should this include the preexisting `pItem` variable? I've left it for the moment as it'll take a full refactor of `BuyMenu.cpp` to remove.

garethyr replied:
Yes, this should include pItem, but as you say, that can wait til a full refactor happens.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1208,20 +1208,45 @@ void BuyMenuGUI::Update()
+            if (pItemAsCraft) {
+                int craftMaxPassengers = pItemAsCraft->GetMaxPassengers();
+                int craftMaxMass = static_cast<int>(pItemAsCraft->GetMaxMass()) - static_cast<int>(pItemAsCraft->GetMass());
+                description += "\nMax Mass: " + std::to_string(craftMaxMass) + "\nMax Passengers: " + std::to_string(craftMaxPassengers);
```
I wonder if it'd be better to have an extra newline before MaxMass. I guess I should actually look at it before I say anything haha.

traunts replied:
I've uploaded screenshots below. Let me know what you think.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1208,20 +1208,45 @@ void BuyMenuGUI::Update()
+                    if (passengerSlotsTaken > 1) {
+                        description += "\nPassenger Slots: " + std::to_string(passengerSlotsTaken);
+                    }
+                }else {
```
missing a space before your else. Applies below too.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1208,20 +1208,45 @@ void BuyMenuGUI::Update()
+            }
         }
 
+        // Show popup info box next to selected item, but only if it has a description
```
I think this is junk?

traunts replied:
This was a legacy comment, I'll remove it.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1353,10 +1378,36 @@ void BuyMenuGUI::Update()
+                    description += "\nPassenger Slots: " + std::to_string(passengerSlotsTaken);
+                }
+            }
+            else {
```
This shouldn't be on a new line.

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1353,10 +1378,36 @@ void BuyMenuGUI::Update()
+                if (pItemAsMO) {
+                    floatStream << std::fixed << std::setprecision(1) << pItemAsMO->GetMass();
+                    // If the item's mass is an integer, remove trailing zero and decimal point.
+                    description += "\nMass: " + ((floatStream.str().back() == '0') ? floatStream.str().substr(0, floatStream.str().length() - 2) : floatStream.str());
```
Given how often you do this stringification, I think it'd be a good idea to add a method to RTETools that lets you pass in a number and get it stringified and trimmed/expanded to some passed-in precision. That way your code gets cleaner and other places have a consistent way of doing it that they can reuse.

traunts replied:
Now implemented as an external function, can round to variable precision. This'll require it's own review as well.

garethyr reviewed (changes requested):
Good work, I've left some comments for things that need changing.
Also, would you mind taking a few screenshots of it? I'll check out the branch and test it myself at some point, but this would be helpful in the meantime.

traunts commented:
Pictured: Screenshots of the tooltips, in both the cart and catalogue focus.
![Catalogue-CoalitionSoldierHeavy](https://user-images.githubusercontent.com/73866773/104885487-0c4a9100-59b8-11eb-9aa7-c4df33487c73.png)
![Catalogue-Detonator](https://user-images.githubusercontent.com/73866773/104885488-0ce32780-59b8-11eb-805e-85608b2b12ff.png)
![Catalogue-DropPod](https://user-images.githubusercontent.com/73866773/104885493-0d7bbe00-59b8-11eb-870e-0a4565828194.png)
![Catalogue-DropshipMk1](https://user-images.githubusercontent.com/73866773/104885497-0e145480-59b8-11eb-8f34-b276268d3e1e.png)
![Catalogue-GatlingDrone](https://user-images.githubusercontent.com/73866773/104885499-0eaceb00-59b8-11eb-9200-51c924811951.png)
![Catalogue-RocketMk2](https://user-images.githubusercontent.com/73866773/104885502-0f458180-59b8-11eb-95a3-07cfbc0e2925.png)
![Catalogue-Shotgun](https://user-images.githubusercontent.com/73866773/104885505-0f458180-59b8-11eb-8c9c-b8cd826189d4.png)
![Cart-CombatRobot](https://user-images.githubusercontent.com/73866773/104885507-0fde1800-59b8-11eb-9573-9137af39cdac.png)
![Cart-CulledClone](https://user-images.githubusercontent.com/73866773/104885509-1076ae80-59b8-11eb-8d89-477cf232c465.png)
![Cart-GatlingDrone](https://user-images.githubusercontent.com/73866773/104885512-110f4500-59b8-11eb-8698-dd48eff99ccd.png)
![Catalogue-BattleRifle](https://user-images.githubusercontent.com/73866773/104885513-11a7db80-59b8-11eb-95a2-bca377bcc49b.png)

traunts commented:
A question: how should this apply to presets? Should it show total mass, any details of the craft etc? I initially chose not to include them as they're already set, so the mass and passenger constraints have to be valid anyway.

garethyr commented:
Regarding the screenshots:
1. Looks good but have you tried having 2 \n's? I wonder if it'd look better with an empty line in between the text and the mass stuff. Similarly, I wonder about having the text for that stuff centered or right aligned separate from the other text (if that's possible).
2. When there's no description, to avoid that weird empty line at the top, maybe put some placeholder string "-No Information Found-" or something.

Regarding presets:
I think you're right, the preset presumably has to be valid, so assuming you can't put in an overweight preset (e.g. with the no passenger or mass limit scripts), I think it's totally fine to not have this info on them. If you can put in invalid presets, maybe for those you'll wanna have it show something like `Warning - Over Mass Limit`?

garethyr commented on System/RTETools.cpp:
```diff
@@ -190,4 +190,18 @@ namespace RTE {
+		std::stringstream floatStream;
+		floatStream << std::fixed << std::setprecision(precision) << input;
+		std::string outputString = floatStream.str();
+		// Remove trailing zeroes and decimal point.
```
I don't think we should remove trailing zeroes here. Just by the technical (i.e. scientific) definition of precision, if I say I want `1.5` with 3 precision, I should get `1.500`. Similarly, 2 precision would be `1.50`, 1 precision would be `1.5`, 0 precision would be `1`.
From a user point of view, this ensures you get exactly what you ask for out of it, instead of sometimes not doing what the method name suggests.

Also, I feel like this might be useful for lua users, wanna expose it in luaman?

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1319,45 +1331,55 @@ void BuyMenuGUI::Update()
+            const Entity *currentItem = pItem->m_pEntity;
+            const Actor *itemAsActor = dynamic_cast<const Actor*>(currentItem);
+            if (itemAsActor) {
+                // If the item's mass is an integer, remove trailing zero and decimal point.
```
Given we have a friendly method that does this, I think there's no need for a description here.
Also, I think it'd probably look fine if it shows `Mass: 50.0 kg` or something, which would allow for the precision changes I mentioned below. To my mind this makes it feel, in a non-technical sense, more precise haha.

traunts replied:
Ah, that was supposed to be gone. Now removed.

traunts replied:
I'll implement this and leave screenshots to compare, I just judged the integer masses slightly easier to read.

traunts commented:
Pictured: Revised screenshots for the BuyMenu Tooltips.
<details>
  <summary>Craft</summary>
  
![Catalogue-RocketMk2](https://user-images.githubusercontent.com/73866773/105189215-a5c69e00-5b84-11eb-8d59-4bf7653b85cb.png)
![Catalogue-Droppod](https://user-images.githubusercontent.com/73866773/105189223-a7906180-5b84-11eb-9ef4-6498e79b155c.png)
  
</details>

<details>
  <summary>Weapons</summary>
  
![Catalogue-ARLight](https://user-images.githubusercontent.com/73866773/105189342-cb53a780-5b84-11eb-8bde-b759473fc426.png)
![Catalogue-Shotgun](https://user-images.githubusercontent.com/73866773/105189350-cd1d6b00-5b84-11eb-9c8c-2a53dbc16cb9.png)
  
</details>

<details>
  <summary>Actors</summary>
  
![Catalogue-SoldierLight](https://user-images.githubusercontent.com/73866773/105189392-d6a6d300-5b84-11eb-9ef3-0b331ffbe737.png)
![Catalogue-GatlingTurret](https://user-images.githubusercontent.com/73866773/105189385-d575a600-5b84-11eb-9300-060936fa5fa8.png)
  
</details>

<details>
  <summary>Cart View</summary>
  
![Cart-BattleRifle](https://user-images.githubusercontent.com/73866773/105189471-e9b9a300-5b84-11eb-8b1a-c18bc4dbcc49.png)
![Cart-SMG](https://user-images.githubusercontent.com/73866773/105189476-eb836680-5b84-11eb-8c21-eae68f57dc35.png)
  
</details>

garethyr commented:
Hmmm, looking at the new screenshots, I don't mind showing the trailing zero for masses like 10.0 kg personally. However, maybe it'd be worth sharing these screenshots in #game-discussion (or somewhere else if there's a more appropriate channel) and seeing what people think.

If we do get rid of the trailing 0, I think it'd be better to either keep that code as a separate method, or only do it when an optional flag (that defaults to off) is set - some `bool trimTrailingZeroes`.

Also, do crafts not show that trailing 0?

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -1319,45 +1335,58 @@ void BuyMenuGUI::Update()
-        if (pItem && pItem->m_pEntity && !pItem->m_pEntity->GetDescription().empty())
-        {
+        if (pItem && pItem->m_pEntity) {
+            //description = (pItem->m_pEntity->GetDescription() != "") ? pItem->m_pEntity->GetDescription() : "-No Information Found-";
```
Seems like some junk code here.

traunts replied:
Yep, those comments were just test stuff, removed now. Screenshots above should give an example of it.

traunts commented:
> Hmmm, looking at the new screenshots, I don't mind showing the trailing zero for masses like 10.0 kg personally, and I think the extra new line looks better (likewise, adding kg after mass). However, maybe it'd be worth sharing these screenshots in #game-discussion (or somewhere else if there's a more appropriate channel) and seeing what people think.
> 
> If we do want to get rid of the trailing 0, I think it'd be better to either keep that code as a separate method. Alternatively, it could be part of that method but only happen when an optional flag (that defaults to off) is set - some `bool trimTrailingZeroes` method param.
> 
> Also, do crafts not show that trailing 0?

Craft mass limits aren't calculated the same way, that's just done by comparing the difference between the craft's mass and it's maximum carriable mass and casting it to an `int`, purely because I'd not encountered float-masses for craft. I can rework it to use the same method as the other tooltips and display as a float if you'd like.

garethyr commented:
Good work, I'm happy with this :)

traunts commented:
I've run into an issue I feared might arise here. The rounding from the `std::precision` stream arg will round down for remainders `0 < r <=0.5` and round up for `0.5 < r <1`. As an example, testing this with a modded item weighing 10.00 < m <= 10.05 kg and a craft with a max mass of 10 kg, we see the item's mass displayed in the buymenu as "10.0 kg", and when added to the cart the mass limit counter will read "10/10" (as it's cast to ints to display it) , however the order is not purchasable cos the check to see if it's valid compares the two as floats.

The biggest problem is it gives a inaccurate reading, ie the player is told that X of this item should weigh a set amount, but the true mass of the order is greater by less than 0.1 kg, so the BuyMenu rejects it.

I can see ways to solve this, such as either adding a tolerance to the mass check in the `BuyMenuGUI` method `TryPurchase( )`, so that if the mass exceeds the maximum by less than 0.05 kg then the order is still valid, or by recasting the rounded string as a float, and if the original mass is greater then add (1/ 10^p), where p is the precision we round to, to the tooltip display mass.

Any input would be appreciated. It only affects items who's mass is specified to more than 1 dp (looking at you, Culled Clones) and it's a distinct edge case that's unlikely to be encountered without trying, but it could still be an issue.

garethyr commented:
Hmmm, a few thoughts:

i. It would probably be sensible to make the mass limit have the same precision as shown in descriptions, i.e. 1 decimal point, to keep things consistent.

ii. I may be wrong, but it seems to me that this could be solved with a parameter to the rtetools method and a bit of math:

1. Add a rounding mode flag that takes an enum or an int if you wanna be lazy. The values would be 0 - naturalRounding (or some other name, but basically what you have now), 1 - alwaysFloor, 2 - alwaysCeil.

2. If rounding mode is 1 or 2, you do the simple math of `result = floorOrCeil(input * (10 ^ precision)) / (10 ^ precision)`. This shifts your decimal place such that your desired precision will be to the left of it, floors or ceils it, then shifts back. You'd probably then run that floored/ceiled number through the precision trimmer to get the appropriate output So, to run through a few examples, where `n` is the number and `p` is the desired precision:
`n0.05 p1` -> `ceil(0.05 * 10) / 10 -> 0.1` or `floor(0.05 * 10) / 10 -> 0.0`
`n0.758349 p3` -> `ceil(0.758349 * 1000) / 1000 -> 0.759` or `floor(0.758349 * 1000) / 1000 -> 0.758`

traunts commented:
If you've followed on Discord, the scope of this has now updated. In fixing the specific cases for rounding errors, I've also updated the Mass Label to match current precision, which lead reworking the `Actor` property, `MaxMass` into `MaxInventoryMass` that no longer includes the Actor's mass in the calculation. This causes issues with craft with carrying capacities, as those were set in `.ini` as their maximum carriable mass + the total mass of the craft itself (parent & all attachables). As such, this now requires a companion [#Data branch](/cortex-command-community/Cortex-Command-Community-Project-Data/tree/GI-251-ACraft-MaxMass-update) where all the uses of MassMax will eventually need to be replaced, and any Lua using it's old calculation will need to be updated. Has to be reviewed manually.

traunts commented:
Note: this still contains the backwards compatibility of MaxMass in both Lua and `ini`, Can easily be removed with approval.

garethyr commented on CHANGELOG.md:
```diff
@@ -150,6 +152,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 - `GetParent` will now always return null for objects with no parents, instead of returning the self object for things that weren't `Attachables`. This makes things more consistent and reasonable throughout and will rarely, if ever, cause Lua problems.
 
+- BuyMenu tooltips now display item info as well as a description. This includes the maximum carriable mass and passengers for Craft, mass and passenger slots required for Actors, and mass for other Moveable Objects.
```
Formatting tweaks needed. For anything that's an actual property or class, we write it out as such and surround it in ``s. I would also suggest re-ordering things a little bit too.
I.e.
- The `Actor` property `MaxMass` now no longer includes the `Mass` of the `Actor`, and has been renamed to `MaxInventoryMass` for clarity. In mods, this is most important for `ACraft`, which will now need their total `Mass` subtracted from the value.

- `BuyMenu` tooltips now display useful item info as well as their description. This includes `MaxInventoryMass` and `MaxPassengers` for `ACraft`, `Mass` and `PassengerSlots` required for `Actors`, and `Mass` for other `MoveableObjects`.

garethyr commented on Entities/Actor.cpp:
```diff
@@ -385,8 +383,8 @@ int Actor::ReadProperty(const std::string_view &propName, Reader &reader)
     }
-    else if (propName == "MaxMass")
-        reader >> m_MaxMass;
+    else if (propName == "MaxMass" || propName == "MaxInventoryMass")
```
Please consider this the go-ahead to kill old propname

garethyr commented on Entities/Actor.h:
```diff
@@ -914,16 +914,12 @@ ClassInfoGetters
+	/// <summary>
+	/// Returns the maximum total mass this Actor can carry in it's inventory.
+	/// </summary>
+	/// <returns> The maximum carriable mass of this Actor.</returns>
```
Should be `its` instead of `it's` in the summary
Also, remove the space at the start of the returns comment

garethyr commented on Entities/Actor.h:
```diff
@@ -1351,9 +1347,9 @@ ClassInfoGetters
-    float m_MaxMass;
+    // The max combined weight of items in this Actor's inventory.
+    // If this is -1 there's no limit.
+    float m_MaxInventoryMass;
```
Might as well redo this member variable comment in our modern standard, and tweak it to be a little better
`float m_MaxInventoryMass; //!< The mass limit for this Actor's inventory. -1 means there's no limit.`

garethyr commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,28 @@ namespace RTE {
+			floatStream << std::fixed << std::setprecision(precision) << input;
+			return floatStream.str();
+		} else {
+			RTEAssert(powf(10.0F, static_cast<float>(precision)) < std::numeric_limits<float>::max(), "PRECISION SET GREATER THAN EXPONENT MAX!");
```
lol no need to capitalize all these asserts
Also, I don't really like having all these asserts redoing similar calculations. Extract the `powf(10.0F, static_cast<float>(precision))` into a variable and reuse it.

garethyr commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,28 @@ namespace RTE {
+			RTEAssert(powf(10.0F, static_cast<float>(precision)) > 0, "NEGATIVE PRECISION WILL YEILD DIVIDE BY ZERO ERROR!");
+			RTEAssert(input < (std::numeric_limits<float>::max() / powf(10.0F, static_cast<float>(precision))), "VALUE WILL EXCEED NUMERIC LIMITS WITH PRECISION " + std::to_string(precision));
+			if (roundingMode == 1) {
+				float roundingBuffer = floor(input * powf(10.0F, static_cast<float>(precision))) / powf(10.0F, static_cast<float>(precision));
```
You may be able to compress this by storing the rounding mode function (floor or ceil) in a variable or using it in a ternary. Basically, would be nice to remove some duplicated code.

traunts replied:
Done, if you don't like the switch-case let me know.

garethyr commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,28 @@ namespace RTE {
+				float roundingBuffer = ceil(input * powf(10.0F, static_cast<float>(precision))) / powf(10.0F, static_cast<float>(precision));
+				return RoundFloatToPrecision(roundingBuffer, precision);
+			} else {
+				RTEAssert(false, "Error in RoundFloatToPrecision: INVALID ROUNDING MODE");
```
You wanna use RTEAbort for this. No need for a return after it, and it's obviously cleaner

traunts replied:
Aye, that was just VS suggesting I avoid possible completion without return conditions.

garethyr commented on System/RTETools.h:
```diff
@@ -216,6 +216,17 @@ namespace RTE {
+	/// Rounds a float to a set fixed point precision (digits after decimal point) with option to always ceil or always floor the remainder.
+	/// </summary>
+	/// <param name="inputFloat">The input float to round.</param>
+	/// <param name="roundingMode">(0, 1, 2) What method of rounding to use for the result. 
```
I don't think you really need to use <para> here, but by all means feel free to. I don't think it's used elsewhere, but maybe it should be lol. That said, you're missing it for roundingMode 2, and I'd suggest putting a `-` between the number and the text, like `<para>0 (default option) - System default.</para>`

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -740,30 +740,22 @@ float BuyMenuGUI::GetTotalOrderCost()
-		totalMass += dynamic_cast<const MOSprite *>((*itr)->m_pEntity)->GetMass();
-
-	totalMass += GetCraftMass();
+	for (vector<GUIListPanel::Item*>::iterator itr = m_pCartList->GetItemList()->begin(); itr != m_pCartList->GetItemList()->end(); ++itr) {
```
Would be nice to replace this with a range for loop if you can
```cpp
for (const GUIListPanel:Item *cartItem : m_pCartList->GetItemList()) {
   totalMass += cartItem->GetMass();
}
```

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -2233,24 +2253,58 @@ void BuyMenuGUI::AddPresetsToItemList()
-    std::snprintf(newText, sizeof(newText), "Cost: %.0f/%.0f", GetTotalOrderCost(), g_ActivityMan.GetActivity()->GetTeamFunds(whichTeam));
-    m_pCostLabel->SetText(newText);
+void BuyMenuGUI::UpdateTotalCostLabel(int whichTeam) {
+	// Total order cost is always rounded up from any remainder.
```
No need for this comment really

garethyr commented on Menus/BuyMenuGUI.cpp:
```diff
@@ -2233,24 +2253,58 @@ void BuyMenuGUI::AddPresetsToItemList()
+			display += " / " + std::to_string(pCraft->GetMaxPassengers());
+		}
+	} else {
+		display = "NO SPACE";
```
Should this say `NO PASSENGER SLOTS` or `NO PASSENGER SPACE`? Maybe it's fine as-is.

traunts replied:
So this was actually what I had initially, and it looked far better, _but_ ;
The text overlaps the rest of the menu. 
![image](https://user-images.githubusercontent.com/73866773/111907000-5a0c7400-8aa3-11eb-808d-656ab74fca5c.png)
It would look great, but I'm guessing it's gonna be a ton more work to just make the GUI resize as required?
I'd prefer it to have the word "`PASSENGER`" in there as well, what are your thoughts?

traunts replied:
I'm leaving this for now, with a personal note that if the GUI gets updated to factor it in.

garethyr reviewed (changes requested):
Hey Stewie, sorry it took me a while to get to this, left some comments with changes that need to be made. Nothing too huge

traunts commented:
Per suggestion, I've currently solved a problem of poorly represented `Mass` values for `Attachables` that shouldn't add to their parent by having the reader set them to `0` before adding them to their parent. This includes things like flashes for `AEmitters` and `HDFirearms`.
These changes aren't committed however, will likely be part of cleanup in another attachable branch. They're also accompanied by changes to the corresponding properties in the data branch, cortex-command-community/Cortex-Command-Community-Project-Data#91.

MaximDude commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,36 @@ namespace RTE {
 	}
-}
\ No newline at end of file
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Newline before this.

MaximDude commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,36 @@ namespace RTE {
+			RTEAssert(precisionMagnitude < std::numeric_limits<float>::max(), "Precision set greater than able to display (exponent too high)!");
+			RTEAssert(precisionMagnitude > 0, "Negative precision will yeild divide by zero error!");
+			RTEAssert(input < (std::numeric_limits<float>::max() / precisionMagnitude), "Value will exceed numeric limits with precision " + std::to_string(precision));
+			
```
Appears you have some trailing whitespace here.

MaximDude commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,36 @@ namespace RTE {
+		} else {
+			float precisionMagnitude = powf(10.0F, static_cast<float>(precision));
+			RTEAssert(precisionMagnitude < std::numeric_limits<float>::max(), "Precision set greater than able to display (exponent too high)!");
+			RTEAssert(precisionMagnitude > 0, "Negative precision will yeild divide by zero error!");
```
`yeild` -> `yield`.

MaximDude commented on System/RTETools.cpp:
```diff
@@ -107,4 +107,36 @@ namespace RTE {
+					RTEAbort("Error in RoundFloatToPrecision: INVALID ROUNDING MODE");
+					break;
+			}
+
```
Remove whitespace here, on 138 and 141.
Keep our files nice and tidy.

MaximDude commented on System/RTETools.h:
```diff
@@ -216,6 +216,17 @@ namespace RTE {
+	/// <summary>
+	/// Rounds a float to a set fixed point precision (digits after decimal point) with option to always ceil or always floor the remainder.
+	/// </summary>
+	/// <param name="inputFloat">- The input float to round.</param>
```
Remove `- ` before param comment.

MaximDude commented on System/RTETools.h:
```diff
@@ -216,6 +216,17 @@ namespace RTE {
+	/// </summary>
+	/// <param name="inputFloat">- The input float to round.</param>
+	/// <param name="roundingMode">- (0, 1, 2) What method of rounding to use for the result. 
+	/// <para>0 (default option) - System default.</para>
```
Man this looks horrible.
Just do 
`/// <param name="roundingMode">Method of rounding to use. 0 for system default, 1 for floored remainder, 2 for ceiled remainder.</param>`.

MaximDude commented on Entities/HDFirearm.cpp:
```diff
@@ -336,8 +336,10 @@ void HDFirearm::SetFlash(Attachable *newFlash) {
         if (m_pFlash && m_pFlash->IsAttached()) { RemoveAttachable(m_pFlash); }
         m_pFlash = nullptr;
     } else {
+		newFlash->SetMass(0.0F);
```
`m_pFlash->SetMass(0);` further down with the rest of the setters.

MaximDude commented on Entities/Actor.h:
```diff
@@ -914,16 +914,12 @@ ClassInfoGetters
-// Return value:    The max recommend total mass for this Actor
-
-	float GetMaxMass() const { return m_MaxMass; }
+	float GetMaxInventoryMass() const { return m_MaxInventoryMass; }
```
Remove whitespace between declaration and comment when reformatting otherwise VS doesn't associate them and hovering over a method call doesn't show the description.
Same thing happens with the old comment style be we don't care so no need to fix those.

MaximDude commented on Entities/AEmitter.cpp:
```diff
@@ -365,6 +365,8 @@ void AEmitter::SetFlash(Attachable *newFlash) {
         if (m_pFlash && m_pFlash->IsAttached()) { RemoveAttachable(m_pFlash); }
         m_pFlash = nullptr;
     } else {
+		newFlash->SetMass(0.0F);
```
`m_pFlash->SetMass(0);` further down with the rest of the setters.

MaximDude commented on CHANGELOG.md:
```diff
@@ -168,6 +170,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - `Attachable` terrain collisions will now propagate to any child `Attachables` on them. This means that `Attachables` will not collide with terrain, even if set to, if they're attached to a parent that doesn't collide with terrain.  
 	This means that the `attachable.CollidesWithTerrainWhileAttached` value may not represent the true state of things, you should instead use `attachable.CanCollideWithTerrain` to determine whether a given `Attachable` can collide with terrain.
+	
+- The `Actor` property`MaxMass` now no longer includes the `Mass` of the `Actor`, and has been renamed to `MaxInventoryMass` for clarity. In mods, this is most important for `ACraft`, which will now need their total `Mass` subtracted from the old value. 
```
Space between `property` and `MaxMass`.

MaximDude commented on CHANGELOG.md:
```diff
@@ -95,6 +95,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 - Added `OnStride` special Lua function for `AHumans` that is called whenever they stride (i.e. when their `StrideSound` is played). Like playing `StrideSound`, this does not happen when the AHuman is climbing.
 
+- Added the helper function `RoundFloatToPrecision( )` to RTETools and exposed to Lua. Utility function to round and format floating point numbers for display in strings. 
```
No need to mention adding to `RTETools`, and since it's exposed to Lua you should explain how to call it.

 
Added Lua utility function `RoundFloatToPrecision` to round and format floating point numbers for display in strings.
```
RoundFloatToPrecision(floatValue, digitsPastDecimal, roundingMode) -- Rounding mode 0 for system default, 1 for floored remainder, 2 for celied remainder
```
	.

traunts commented on Entities/MovableObject.cpp:
```diff
@@ -265,13 +265,9 @@ int MovableObject::Create(const MovableObject &reference)
+	if (propName == "Mass") {
 		reader >> m_Mass;
-		if (m_Mass == 0)
-			m_Mass = 0.0001;
```
As I said in discord, I'll add checks wherever there's division done with it, and have it default to 0.0001 as before. 
I've just made my first pass with it, I'll keep my eye out for other instances.

MaximDude reviewed (dismissed):
Don't really like using floats as bools for evaluation but whatever, 0 can't screw up and become a 0.00000001.
Might want to look for a better solution later:tm:.

garethyr reviewed (approved):
I made a few minor tweaks. Beyond that, it looks good, great work!
